bokeh/util/compiler.py

```diff
@@ -0,0 +1,282 @@
+from __future__ import absolute_import
+
+import logging
+logger = logging.getLogger(__name__)
+
+import six
+import json
+import inspect
+import hashlib
+from os.path import basename, dirname, join, abspath, relpath, exists
+from subprocess import Popen, PIPE
+
+from ..model import Model
+from ..settings import settings
+from .string import snakify
+from .paths import bokehjsdir
+
+# XXX: this is the same as bokehjs/src/js/plugin-prelude.js
+_plugin_prelude = \
+"""
+(function outer(modules, cache, entry) {
+  if (typeof Bokeh !== "undefined") {
+    for (var name in modules) {
+      Bokeh.require.modules[name] = modules[name];
+    }
+
+    for (var i = 0; i < entry.length; i++) {
+        Bokeh.Models.register_locations(Bokeh.require(entry[i]));
+    }
+  } else {
+    throw new Error("Cannot find Bokeh. You have to load it prior to loading plugins.");
+  }
+})
+"""
+
+_plugin_template = \
+"""
+%(prelude)s
+({
+  "custom/main": [function(require, module, exports) {
+    module.exports = { %(exports)s };
+  }, {}],
+  %(modules)s
+}, {}, ["custom/main"]);
+"""
+
+_style_template = \
+"""
+(function() {
+  var head = document.getElementsByTagName('head')[0];
+  var style = document.createElement('style');
+  style.type = 'text/css';
+  var css = %(css)s;
+  if (style.styleSheet) {
+    style.styleSheet.cssText = css;
+  } else {
+    style.appendChild(document.createTextNode(css));
+  }
+  head.appendChild(style);
+}());
+"""
+
+_export_template = \
+"""%(name)s: require("%(module)s")"""
+
+_module_template = \
+""""%(module)s": [function(require, module, exports) {\n%(code)s\n}, %(deps)s]"""
+
+class AttrDict(dict):
+    def __getattr__(self, key):
+        return self[key]
+
+class CompilationError(RuntimeError):
+
+    def __init__(self, error):
+        super(CompilationError, self).__init__()
+        self.line = error.get("line")
+        self.column = error.get("column")
+        self.message = error.get("message")
+        self.text = error.get("text")
+        self.annotated = error.get("annotated")
+
+    def __str__(self):
+        return self.text
+
+bokehjs_dir = settings.bokehjsdir()
+
+def nodejs_compile(code, lang="javascript", file=None):
+    compilejs_script = join(bokehjs_dir, "js", "compile.js")
+
+    try:
+        proc = Popen(["nodejs", compilejs_script], stdout=PIPE, stderr=PIPE, stdin=PIPE)
```

Comment:
@canavandl, at least on Ubuntu, the default is `nodejs` and `node` is a symlink. I think conda package for node.js should create both. Though, I think I will add detection code to bokeh anyway.